Sources/History.swift

```diff
@@ -0,0 +1,369 @@
+    ///
+    /// - parameter query: The searched query. By convention, an empty search matches the entire history
+    ///                    (up to `options.maxHits`).
+    /// - parameter options: Search options.
+    /// - returns: A list of matching hits in the history. They are ordered by recency (most recent first).
+    ///
+    @objc public func search(query: SearchParameters, options: HistorySearchOptions = HistorySearchOptions()) -> [HistoryHit] {
+        var hits: [HistoryHit] = []
+        let text = query.query ?? ""
+        let searchedText = normalize(text)
+        // Only search for the text at the beginning of a word.
+        let searchedPattern = "\\b" + searchedText
+
+        for query in queries {
+            // Stop when `maxHits` is reached.
+            if hits.count >= options.maxHits {
+                break
+            }
+            let queryString = query as! String
+            
+            var highlightedText: String
+            if searchedText.isEmpty { // empty search: everything matches, no highlighting necessary
+                highlightedText = queryString
+            } else { // non-empty search
+                if let foundRange = queryString.range(of: searchedPattern, options: .regularExpression) {
+                    // NOTE: We don't handle multiple occurrences as they are assumed to be rare.
+                    // Also, highlighting would be weird in a suggestion context.
+                  highlightedText = String(queryString[..<foundRange.lowerBound]) + options.highlightPreTag + String(queryString[foundRange]) + options.highlightPostTag + String(queryString[foundRange.upperBound...])
+                } else {
+                    continue
+                }
+            }
+            // Add hit.
+            let params = SearchParameters()
+            params.query = queryString
+            hits.append(HistoryHit(params: params, highlightedText: highlightedText))
+        }
+        return hits
+    }
+    
+    /// Add a new query to the history.
+    ///
+    /// Deduplication will occur:
+    ///
+    /// 1. If the new query is a prefix of an existing query not stopping at a word boundary, it is discarded.
+    ///   For example, if the history contains "star wars", "star war" will not be added, but "star" will be.
+    /// 2. If an existing query is a prefix of the new query not stopping at a word boundary, the existing query is
+    ///   replaced by the new one. For example, if the history contains "star war", "star wars" will replace it,
+    ///   but "star war fan" will add a new entry.
+    /// 3. If the same query already exists, it is updated.
+    ///
+    /// When the new query does make it into the history, it automatically becomes the most recent entry, possibly
+    /// evicting the less recent entry (LRU principle).
+    ///
+    /// - parameter params: The search parameters of the query to add.
+    ///
+    @objc public func add(_ params: SearchParameters) {
```

Comment:
this function needs to add more smaller helper functions so that it is easier to read and understand. we can discuss offline

Reply:
Let's chat offline